Sources/ConfigCat/Mutex.swift

```diff
@@ -0,0 +1,39 @@
+#if os(Linux)
+import Glibc
+#else
+import Darwin
+#endif
+
+import Foundation
+
+class Mutex {
+    private let mutex: UnsafeMutablePointer<pthread_mutex_t> = UnsafeMutablePointer.allocate(capacity: 1)
```

Comment:
What is the rationale behind using `pthread_mutex` vs. other primitives such as `DispatchSemaphore`?

Reply:
I did some digging in the synchronisation topic and I found out that it's not recommended to use `DispatchSemaphore` in mutex-like scenarios. Mainly because semaphores are not tied to a thread (like mutexes), so the QoS mechanism could not apply thread priority logic to threads that locks on a semaphore, because there's no way to tell which thread owns it. Chances are low but this could end in thread starvation and deadlocks. 
For the same reason it's not recommended to use them for synchronising async callbacks as we did in `AsyncResult`, so it was an other indicator to get rid of it. :)

Reply:
@z4kn4fein  A side note to the `Mutex` implementation:

If calls to lock/unlock are not balanced, `pthread_mutex_destroy` will not be successful upon `Mutex.deinit`. That is, if `pthread_mutex_destroy` returns a value other than zero, it failed. This can happen if one locked a mutex, but then forgot to unlock before the mutex was deallocated. While this does not seem to be the issue here as `Mutex.unlock` is called via `defer`, it is perhaps worth mentioning with a comment that particular care has to be taken with respect to the life cycle of `Mutex`. Since the return value of `pthread_mutex_destroy` is not handled now, perhaps one could assert that destroy indeed did work as intended in unoptimized code (e.g. via `assertionFailure`)?

Reply:
That's a good point, thanks! I've decorated the mutex with asserts.